src/main/scala/org/apache/spark/rdd/OrderedRDD.scala

```diff
@@ -8,105 +8,112 @@ import scala.collection.mutable.ArrayBuffer
 import scala.reflect.ClassTag
 import scala.util.hashing._
 
+case class PartitionKeyInfo[T](
+  partIndex: Int,
+  sortedness: Int,
+  min: T,
+  max: T)
+
+object PartitionKeyInfo {
+  final val UNSORTED = 0
+  final val TSORTED = 1
+  final val KSORTED = 2
+
+  def apply[T, K](partIndex: Int, projectKey: (K) => T, it: Iterator[K])(implicit tOrd: Ordering[T], kOrd: Ordering[K]): PartitionKeyInfo[T] = {
+    assert(it.hasNext)
+
+    val k0 = it.next()
+    val t0 = projectKey(k0)
+
+    var minT = t0
+    var maxT = t0
+    var sortedness = KSORTED
+    var prevK = k0
+    var prevT = t0
+
+    while (it.hasNext) {
+      val k = it.next()
+      val t = projectKey(k)
+
+      if (tOrd.lt(prevT, t))
+        sortedness = UNSORTED
+      else if (kOrd.lt(prevK, k))
+        sortedness = sortedness.min(TSORTED)
+
+      if (tOrd.lt(t, minT))
+        minT = t
+      if (tOrd.gt(t, maxT))
+        maxT = t
+
+      prevK = k
+      prevT = t
+    }
+
+    PartitionKeyInfo(partIndex, sortedness, minT, maxT)
+  }
+}
+
 object OrderedRDD {
 
   def empty[T, K, V](sc: SparkContext, projectKey: (K) => T)(implicit tOrd: Ordering[T], kOrd: Ordering[K], tct: ClassTag[T],
     kct: ClassTag[K]): OrderedRDD[T, K, V] = new OrderedRDD[T, K, V](sc.emptyRDD[(K, V)], OrderedPartitioner.empty[T, K](projectKey))
 
-  def apply[T, K, V](rdd: RDD[(K, V)], projectKey: (K) => T, reducedRDD: Option[RDD[K]] = None)
+  def apply[T, K, V](rdd: RDD[(K, V)], projectKey: (K) => T, fastKeys: Option[RDD[K]] = None)
```

Comment:
what happened to checking if the RDD is an ordered RDD or has an ordered partitioner?